### PR TITLE
Added couchdb and redis. Added Docker. Updated to Express 4. Updated dependencies.

### PR DESCRIPTION
**Pull request by maysale01**
To solve the caching issues I added couchdb for storing the packages from npm to disk and redis for caching the search terms in memory. This is a pretty naive implementation at the moment, didn't have much time to refactor. Let me know if you want to push this into Amazon ECS, I can take a go at it.

### CODE REVIEW

File: src/lib/search.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+        console.log('Creating registry database');
+        db.create();
+        callback(null);
+    });
+}
+
+// Get all the mongoose packages from mongoose and load them into couchDB
+function populateDB(callback) {
+    npm.commands.search(['mongoose'], function(error, payload) {
```
Comment (vkarpov15) [inline src/lib/search.js]: One of the big problems is that the memory overhead of this command is huge, this op by itself is enough to crash the process on an EC2 t2.micro. I'm really interested to see if there's any way we can convert this to a stream so we're not dependent on having enough memory to store all the data that npm spits out at once.

**Comment by vkarpov15** [comment]
Beside above comment, if I'm gonna be maintaining this codebase, I have zero interest in learning how to use+maintain CouchDB. Would much rather switch over to mongodb or rethinkdb or datomic if we're gonna be going that route. Also, our fundamental problem from the above comment is that we need to hold all the mongoose packages in memory at once, which makes us need disproportionately powerful hardware. Using a database helps, but that `npm.commands.search()` call is the real problem we need to solve, and I don't understand the npm API well enough to resolve that.

**Comment by maysale01** [comment]
Well, unless you're planning on expanding the functionality of
'mongooseplugins', might as well just make it a single page javascript
application. This is my first experience with CouchDB, but it seems pretty
straightforward and Docker/ECS abstract away any setup pain. Since npm runs
on CouchDB, we can just replicate the NPM registry to the local registry
and it will keep up to date with existing changes.
https://docs.npmjs.com/misc/registry

I didn't have the time to get into this yet. Also, the way the current
filtering works it loads all the records into memory, this needs to be
updated to use partial matching. It already kinda works using 'startkey',
but it has some weird results.

**Comment by vkarpov15** [comment]
Yeah but then we'd have to set up CouchDB to replicate from npm and that sounds like operational headache. Maintaining the plugins site has really not been high up on my priorities list, so I'd rather reduce the amount of babysitting it needs. Perhaps we could just use [JSONStream](http://blog.modulus.io/npm-install-jsonstream) to shove the data into a mongodb or rethinkdb instance?

**Comment by maysale01** [comment]
Actually, CouchDB makes it pretty easy, with Cradle you'd do something like
this:

db.replicate({"source":"https://skimdb.npmjs.com/registry","target":"
http://couchdb:5984/registry", "continuous": true})

CouchDB will poll the source for updates and maintain replication. Might
seem like overkill right now but it gives a lot of flexibility going
forward if you're to expand the functionality of the plugins site.

**Comment by vkarpov15** [comment]
Good suggestion. How does this work performance-wise - does searching for 'mongoose' take 20+ seconds like it does for using `npm.commands.search()`?

**Comment by maysale01** [comment]
No idea to be honest, but I'm assuming you can configure the npm API to use
the local repository and just run npm.commands.search(['mongoose', term]).
I'll experiment with it and see what works best.

**Comment by aheckmann** [comment]
@vkarpov15 do you want me to move this repo somewhere else?

**Comment by vkarpov15** [comment]
@aheckmann you can transfer it over to me if you don't want to get notifications anymore

**Comment by aheckmann** [comment]
made the request

**Comment by vkarpov15** [comment]
Approved. Thanks Aaron, hope you're doing well. I'm gonna be moving to the bay area soon - want to grab a drink when I'm out there?

**Comment by aheckmann** [comment]
sounds good!

**Comment by vkarpov15** [comment]
I'm gonna close this in favor of https://github.com/vkarpov15/mongooseplugins/pull/11 . This PR still doesn't solve the big problem of memory usage with `npm search`, but #11 does. Also gives me a chance to play around with cycle.js, which is pretty sweet :)

**Pull request closed without merging**